auth_pysaml/models/res_users.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (C) 2020 GlodoUK <https://www.glodo.uk>
+# Copyright (C) 2010-2016 XCG Consulting <http://odoo.consulting>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+
+import passlib
+from passlib import pwd
+
+from odoo import SUPERUSER_ID, _, api, fields, models, tools
+from odoo.exceptions import AccessDenied, ValidationError
+
+_logger = logging.getLogger(__name__)
+
+
+class ResUserSaml(models.Model):
+    _name = "res.users.saml"
+
+    user_id = fields.Many2one("res.users", index=True, required=True)
+    saml_provider_id = fields.Many2one(
+        "auth.saml.provider", string="SAML Provider", index=True
+    )
+    saml_uid = fields.Char("SAML User ID", help="SAML Provider user_id", required=True)
+
+    _sql_constraints = [
+        (
+            "uniq_users_saml_provider_saml_uid",
+            "unique(saml_provider_id, saml_uid)",
+            "SAML UID must be unique per provider",
+        )
+    ]
+
+
+class ResUser(models.Model):
+    """
+    Add SAML login capabilities to Odoo users.
+    """
+
+    _inherit = "res.users"
+
+    saml_ids = fields.One2many("res.users.saml", "user_id")
+
+    @api.constrains("password", "saml_ids")
+    def check_no_password_with_saml(self):
+        """Ensure no Odoo user posesses both an SAML user ID and an Odoo
+        password. Except admin which is not constrained by this rule.
+        """
+        if not self._allow_saml_and_password():
+            # Super admin is the only user we allow to have a local password
+            # in the database
+            if self.password and self.id is not SUPERUSER_ID and self.saml_ids:
+                raise ValidationError(
+                    _(
+                        "This database disallows users to "
+                        "have both passwords and SAML IDs. "
+                        "Errors for login %s"
+                    )
+                    % (self.login)
+                )
+
+    def _auth_saml_validate(self, provider_id, token, base_url=None):
+        p = self.env["auth.saml.provider"].sudo().browse(provider_id)
```

Comment:
We should avoid one-letter variables, for readability:

```suggestion
                  provider = self.env["auth.saml.provider"].sudo().browse(provider_id)
```